refactor(typescript): add foundation for incremential TS migration

AllAwesome497 opened the pull request:
<!--
Thanks for wanting to fix something on Sequelize - we already love you!
Please fill in the template below.
If unsure about something, just do as best as you're able.

If your PR only contains changes to documentation, you may skip the template below.
-->

### Pull Request check-list

_Please make sure to review and check all of these items:_

- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Have you added new tests to prevent regressions? N/A, though tests are now run on the compiled code.
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)?

<!-- NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open. -->

### Description of change

Adds a foundation for a migration to TypeScript (not just .d.ts files).  This is intended to be for a gradual migration, and getting this in sooner makes it easier to start converting files to TypeScript once we're ready.  

The primary changes are:
- Added a build script which generates `.js`, `.js.map`, and `.d.ts` files via esbuild and typescript.
- Added eslint rules for TypeScript. 
- Updated existing tests to import from the generated code.
- Converts the logger interface to TS.  This change is intended to ensure that the build script is at least working; the file itself was chosen primarily because it is unlikely to be changed often, and therefore unlikely to run into merge conflicts.

The following (dev) dependencies were added:
- `@types/debug` - Types for the builtin debug package, used by the logger.
- `@typescript-eslint/eslint-plugin` - TypeScript plugin for eslint.
- `@typescript-eslint/parser` - TypeScript parser for eslint.
- `copyfiles` - Used in `build.js` to copy type declaration files from `types/` to `dist/`.
- `esbuild` - Used in `build.js` to transpile the sequelize source code into minified code.  
- `fast-glob` - Used in `build.js` to  locate files that need to be copied / transpiled.
- `module-alias` - Used for testing to alias imports to `sequelize/` to the `dist` directory.  

Additionally, the `strict` rule which required `use strict` at the start of every file was removed.  This is because esbuild will add this in the transpiled files, so it doesn't make sense to require it anymore.

As for why I went with esbuild - 2 primary reasons:
- When I tested it against swc, it was consistently faster (from memory, ~150ms vs ~60ms for esbuild)
- I'm already familiar with esbuild and so it was easier for me to setup.

Note that this does not add support for `.test.ts` files yet - I wanted to limit the complexity of this PR, so I opted to leave that out of this one.  I'll put up a PR with that if this gets approved.

sdepold commented on build.js:
```diff
@@ -0,0 +1,83 @@
+      keepNames: true
+    }),
+
+    promisify(exec)('tsc', {
```
:+1:

sdepold commented on package.json:
```diff
@@ -180,18 +189,18 @@
     "lint": "eslint lib test --quiet",
     "lint-docs": "markdownlint docs",
-    "test-typings": "tsc -b types/tsconfig.json && tsc -b types/test/tsconfig.json",
+    "test-typings": "tsc -b types/tsconfig.json && tsc -b types/test/tsconfig.json && tsc --noEmit --emitDeclarationOnly false",
```
Aaaaaah `--emitDeclarationOnly false` is a pair :D :)

sdepold commented:
Soooo. Can we just merge this or do we have to double check or prepare anything else?

AllAwesome497 commented:
> Soooo. Can we just merge this or do we have to double check or prepare anything else?

Should be good to go!  We'll have to resolve any merge conflicts (which will probably be any test file due to the change in relative paths into `sequelize/*`) but other than that it should be seamless.

sdepold commented:
> > Soooo. Can we just merge this or do we have to double check or prepare anything else?
> 
> Should be good to go! We'll have to resolve any merge conflicts (which will probably be any test file due to the change in relative paths into `sequelize/*`) but other than that it should be seamless.

Ok. And there is no change whatsoever to the release and require / import flow, right?

AllAwesome497 commented:
> Ok. And there is no change whatsoever to the release and require / import flow, right?

For the release flow, `npm publish` should run `npm prepare`.  Under [Life Cycle Scripts](https://docs.npmjs.com/cli/v7/using-npm/scripts#life-cycle-scripts) in the docs for npm:
![image](https://user-images.githubusercontent.com/47748690/139735017-19eedf08-82f1-4575-81a3-5500a7f96922.png)


There shouldn't be any from a user's perspective, though where `require('sequelize/*')` resolves to has changed.
I've verified parts of this via `npm pack` -> `tar -xf sequelize-0.0.0-development.tgz` -> try to require `package/` but never too much testing IMO.

sdepold commented:
> > Ok. And there is no change whatsoever to the release and require / import flow, right?
> 
> For the release flow, `npm publish` should run `npm prepare`. Under [Life Cycle Scripts](https://docs.npmjs.com/cli/v7/using-npm/scripts#life-cycle-scripts) in the docs for npm: ![image](https://user-images.githubusercontent.com/47748690/139735017-19eedf08-82f1-4575-81a3-5500a7f96922.png)
> 
> There shouldn't be any from a user's perspective, though where `require('sequelize/*')` resolves to has changed. I've verified parts of this via `npm pack` -> `tar -xf sequelize-0.0.0-development.tgz` -> try to require `package/`. Never too much testing though, feel free to poke at it through `npm pack`.

once merged, we could try to release an alpha version to npm :)

sdepold commented:
@AllAwesome497 Are you going to enable the test compilation too?

Keimeno commented on lib/utils/logger.ts:
```diff
@@ -0,0 +1,68 @@
+   *
+   * @default 'sequelize'
+   */
+  context?: string;
```
I suppose you removed the `debug` property because it isn't used anywhere in the codebase. @sdepold do you know why we had a `debug` property in the logger config in the first place?

sdepold replied:
Nope no idea

AllAwesome497 commented:
> @AllAwesome497 Are you going to enable the test compilation too?

Yep! just did. 

Tests now run with a require hook (transpiles imports to `.ts` files via esbuild when they're `require`d), so `.test.ts` files will be transpiled via esbuild which will let them be executed, then I also added another `tsconfig.json` for tests which will handle type checking.

I also added a `logger.test.ts` file to ensure that ts tests work properly (and removed the existing tests from the general util tests since they weren't needed anymore).

sdepold commented:
That last commit is a tribute to @tj:

test.js -> test.[tj]s 

😂👋

sdepold commented:
Nice stuff. I'd say let's release a new version of main, merge this guy and attempt another release afterwards.

sdepold commented:
I guess you need to fix the merge conflicts
